Add spanmetricsprocessor logic

albertteoh opened the pull request:
**Description:** Adds spanmetricsprocessor logic as a follow-up to #1917.

**Link to tracking Issue:** #403

**Testing:**
- Unit and integration tested locally (by adding to `cmd/otelcontribcol/components.go`).
- Performance tested (added metrics key cache to improve performance) and added benchmark function.

**Documentation:** Add relevant documentation to exported structs and functions.

Commit:
Add processor logic

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Add license

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Cache metrics dimension key

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Add metrics dimensions cache

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Fix linting issues

Signed-off-by: albertteoh <albert.teoh@logz.io>

albertteoh commented:
The "setup" step failed in CI; I suspect it's flaky. Any chance somebody can rerun this please? Or I'm happy to do it myself if given the permissions to do so; currently the rerun button is disabled for me.

Commit:
Add experimental note in README

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Refactor for readability

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Idiomatic logging

Signed-off-by: albertteoh <albert.teoh@logz.io>

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+func (p *processorImp) buildKey(serviceName string, span pdata.Span) (string, error) {
+	d := p.getDimensions(serviceName, span)
+	p.lock.Lock()
+	lastDimension := p.metricsKeyCache.InsertDimensions(d...)
```
The purpose of metricsKeyCache is not clear to me. Is it done to avoid serializing to json on every buildKey call? Is there any evidence this is less expensive? Also why json? Can we serialize trivially by concatenating dimensions and values sequentially? Would that be less expensive than json and most importantly less expensive than the trie-based cache?

tigrannajaryan replied:
OK, I can see benchmarking of the cache was done, which is good. It is not clear why not use a much simpler serialization instead of json. It may be potentially faster than the cache itself and eliminate the need for the cache.

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+
+// updateCallMetrics increments the call count for the given metric key.
+func (p *processorImp) updateCallMetrics(key string) {
+	p.callSum[key]++
```
Looks like we never retire anything from `callSum`, `latencySum`, etc maps. Will they grow infinitely? Should we reset them periodically with fresh start times?

tigrannajaryan replied:
Sorry, I was not clear. I did not mean growing the value of a particular sum stored in callSum entry. I mean growing the number of entires in the callSum map, so that the map become larger and larger as we observe new keys.

tigrannajaryan replied:
As for resetting the value I think that's the expected behavior for cumulative metrics, see [here] (https://github.com/open-telemetry/opentelemetry-proto/blob/b9f9b7e6d0ecbf00784a02e1b17fd5cc00ee7d73/opentelemetry/proto/metrics/v1/metrics.proto#L269):

>  // CUMULATIVE is an AggregationTemporality for a metric aggregator which
>  // reports changes since a fixed start time. This means that current values
>  // of a CUMULATIVE metric depend on all previous measurements since the
>  // start time. Because of this, the sender is required to retain this state
>  // in some form. **If this state is lost or invalidated, the CUMULATIVE metric**
>  // **values MUST be reset and a new fixed start time following the last**
>  // **reported measurement time sent MUST be used.**

I may be wrong, but that's my understanding of how you can start the counters from 0. @bogdandrutu @jmacd can you confirm?

albertteoh replied:
> I mean growing the number of entires in the callSum map, so that the map become larger and larger as we observe new keys.

Okay, yes indeed the map keys will continue to grow. I'll look into resetting the maps along with resetting start times.

albertteoh replied:
@tigrannajaryan I wrote a small test where on every 10K spans received I try resetting my raw metrics maps and startTime like so:
```
func (p *processorImp) resetRawMetrics() {
	p.counter++
	if p.counter > 10000 {
		p.callSum = make(map[string]int64)
		p.latencySum = make(map[string]float64)
		p.latencyCount = make(map[string]uint64)
		p.latencyBucketCounts = make(map[string][]uint64)
		p.startTime = time.Now()
		p.counter = 0
	}
}
...
		dpLatency := pdata.NewIntHistogramDataPoint()
		dpLatency.SetStartTime(pdata.TimeToUnixNano(p.startTime))
		dpLatency.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
		dpLatency.SetExplicitBounds(p.latencyBounds)
		dpLatency.SetBucketCounts(p.latencyBucketCounts[key])
...
		dpCalls := pdata.NewIntDataPoint()
		dpCalls.SetStartTime(pdata.TimeToUnixNano(p.startTime))
		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
		dpCalls.SetValue(p.callSum[key])
...
```

However, my metric numbers are now jumping around, with some metrics also missing then reappearing again when I query my prometheus server's `/metrics` endpoint. Is there something I'm missing or doing wrong that would prevent Prometheus from stitch together disjoint timeseries?

tigrannajaryan replied:
I don't know much about Prometheus to help. I am fine if you keep the ever-increasing metrics for now and solve this in a future PR.

albertteoh replied:
Okay, I've created a follow-up Issue to address this at a later time: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/2179.

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+func (p *processorImp) collectCallMetrics(ilm *pdata.InstrumentationLibraryMetrics) error {
+	for key := range p.callSum {
+		dpCalls := pdata.NewIntDataPoint()
+		dpCalls.SetTimestamp(pdata.TimeToUnixNano(time.Now()))
```
Do we also need to set the start time? See https://github.com/open-telemetry/opentelemetry-proto/blob/b9f9b7e6d0ecbf00784a02e1b17fd5cc00ee7d73/opentelemetry/proto/metrics/v1/metrics.proto#L265

albertteoh replied:
From what I understand from the link you provided, setting a start time a requirement when periodically resetting cumulative metrics, is that right?

tigrannajaryan replied:
I think so, see my previous reply.

albertteoh commented on processor/spanmetricsprocessor/metricsdimensions/cache.go:
```diff
@@ -0,0 +1,107 @@
+// standard deviation implying a more stable performance footprint.
+// Naturally, this is a trade-off with additional memory used to store the Trie. It is the responsibility of users
+// to configure sensible dimensions with low-cardinality, which is a good practice for metrics in general.
+type Cache struct {
```
Depending on the outcome of your suggested approach, maybe we won't need this cache. But if we still do need the cache, I'll definitely add a diagram.

tigrannajaryan commented on processor/spanmetricsprocessor/metricsdimensions/cache.go:
```diff
@@ -0,0 +1,107 @@
+// CacheNode represents a node in the Cache Trie.
+type CacheNode struct {
+	children  map[string]*CacheNode
+	metricKey *string
```
Anything prevents to use the empty string as the sentinel value? That will avoid unnecessary indirection, which results in the string slice header being allocated on the heap instead of passed by value, thus increasing the number of heap allocations and garbage to be collected.

albertteoh replied:
Interesting, I've learnt something new :). Nope, I think using an empty string would be an acceptable sentinel value too; it wouldn't make sense to have an empty metric key.

As above, this may not even be a concern if we no longer need the cache. :)

tigrannajaryan commented on processor/spanmetricsprocessor/metricsdimensions/cache.go:
```diff
@@ -0,0 +1,107 @@
+// Copyright -c Google LLC
```
If this file in its entirety is copied from elsewhere where originally it has "Copyright -c Google LLC" then you need to keep the notice. If you changed the file you also need to prepend the full OpenTelemetry copyright notice.
If you are the _author_ of the _entire_ code we strongly recommend removing "Copyright -c Google LLC" and use only  OpenTelemetry copyright notice.

But first let's decide whether we need the cache or no at all.

albertteoh commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+	// Binary search to find the latency bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latency)
+
+	p.lock.Lock()
```
Ah, of course. I'm a goose :)

Commit:
Simplify metric key building

Signed-off-by: albertteoh <albert.teoh@logz.io>

Commit:
Fix lint error

Signed-off-by: albertteoh <albert.teoh@logz.io>

albertteoh commented:
@tigrannajaryan I believe I've addressed your comments. Good news also is with your proposed approach, there was a 60% performance improvement based on my benchmarks, compared with the original (before any caching) implementation. It's a 45% performance improvement from my more complicated trie-based cache as well :). Thanks again for the great suggestion!

I've added a comment in code for this, but the most expensive operation is the `WriteString` operation in the StringBuilder, but not sure if we can do much more to optimize this, but open to suggestions.

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -141,15 +198,196 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+			value = *d.Default
+		}
+		// Map the various attribute values to string.
+		if attr, ok := spanAttr.Get(d.Name); ok {
+			switch attr.Type() {
+			case pdata.AttributeValueSTRING:
+				value = attr.StringVal()
+			case pdata.AttributeValueINT:
+				value = strconv.FormatInt(attr.IntVal(), 10)
+			case pdata.AttributeValueDOUBLE:
+				value = strconv.FormatFloat(attr.DoubleVal(), 'f', -1, 64)
+			case pdata.AttributeValueBOOL:
+				value = strconv.FormatBool(attr.BoolVal())
+			case pdata.AttributeValueNULL:
+				value = ""
+			default:
+				p.logger.Warn("Unsupported tag data type", zap.String("data-type", attr.Type().String()))
+				continue
+			}
```
Would `tracetranslator.AttributeValueToString` work instead?

albertteoh replied:
Thanks!

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -141,15 +198,196 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+func (p *processorImp) addDimension(key, value string) {
+	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
+	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
+	p.metricKeyBuilder.WriteString(value)
```
I think we need to delimit the values otherwise we can have key collisions for different dimensions combinations. For example {serviceName="a", operationKey="bc"} will produce the same concatenated key as {serviceName="ab", operationKey="c"} (both result in concatenation of "abc").
Going with a 0 byte as a character should be good enough. 0 byte should not occur in the string value legitimately.
Alternatively, if we think 0 byte can occur in valid attribute values then we will need to escape the string values to ensure the delimiter we use can not be observed in the escaped value alone.

albertteoh replied:
Great catch, @tigrannajaryan. I'll try the 0 byte option.

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -141,15 +198,196 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
+	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
+	p.metricKeyBuilder.WriteString(value)
+	p.dimensionsBuffer[key] = value
```
We do not need to _always_ build `dimensionsBuffer`. We only need to build it if the concatenated key is not found in the cache. Can we delete this line and build the `dimensionsBuffer` only when we discover that the concatenated key is missing in the cache and we need to populate the cache?

albertteoh replied:
Okay, I have modified this to only build the map if there is a cache miss of `metricKey`. My implementation involves some duplicated logic due to the slightly different requirements for building a `metricKey` vs. building a the metric key map. Open to suggestions on how to improve this.

Commit:
metricKey type and use AttributeValueToString

Signed-off-by: albertteoh <albert.teoh@logz.io>

albertteoh commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -141,15 +198,196 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+// any additional dimensions the user has configured.
+// The metric key is a simple concatenation of dimension values.
+func (p *processorImp) buildKey(serviceName string, span pdata.Span) string {
+	// Reset back to a clean state.
+	defer func() {
+		// Compiler will optimize this map clearing operation.
+		// https://github.com/golang/go/blob/master/doc/go1.11.html#L447
+		for k := range p.dimensionsBuffer {
+			delete(p.dimensionsBuffer, k)
+		}
+		p.metricKeyBuilder.Reset()
+	}()
```
Yeah, this is a little premature and I did measure a slight increase in execution time, but it's minimal and probably within reasonable error bounds. Readability is more important.

Commit:
Address review comments

Signed-off-by: albertteoh <albert.teoh@logz.io>

albertteoh commented:
@tigrannajaryan please let me know if there are any other issues that I have not addressed.

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -141,15 +193,187 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+
+func concatDimensionValue(metricKeyBuilder *strings.Builder, value string, prefixSep bool) {
+	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
+	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
```
Nit: we probably can easily have a fairly good idea about the expected size of the final string and can preallocate it. Can be a future improvement.

albertteoh replied:
I did try `stringBuilder.Grow(n)` to initialize the capacity before concatenating strings, but it seemed to increase CPU time. It seems the larger proportion of time taken in `Grow` is the creation of a new slice, whereas growing the existing slice seemed to be cheaper; I don't completely understand why tbh.

I've attached screenshots of the flamegraphs in case you were interested.

![Screen Shot 2021-01-29 at 12 15 33 pm](https://user-images.githubusercontent.com/26584478/106218628-c8fbe800-622b-11eb-9965-0204eb5546fe.png)
![Screen Shot 2021-01-29 at 12 14 57 pm](https://user-images.githubusercontent.com/26584478/106218636-cd280580-622b-11eb-8d2f-8b27c5080527.png)

tigrannajaryan commented on processor/spanmetricsprocessor/processor.go:
```diff
@@ -141,15 +193,187 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
+
+	p.lock.Lock()
+	key := buildKey(serviceName, span, p.dimensions)
```
Nit: I think this line can be before the lock, I do not see it touching any shared data.

tigrannajaryan reviewed (approved):
LGTM. Merging as is, feel free to address the nits in a future PR.

albertteoh commented:
@tigrannajaryan thank you for the thorough review. It has definitely made the quality of this component a lot better, and I've learnt a lot along the way. 🙏 

I'll address the nit relating to the lock. As for the string builder, I'll leave that for later as I'm not sure if that'll be a quick change.
